src/main/java/dev/openfeature/sdk/OpenFeatureAPI.java

```diff
@@ -76,20 +84,49 @@ public EvaluationContext getEvaluationContext() {
      * {@inheritDoc}
      */
     public void setProvider(FeatureProvider provider) {
+        if (provider == null) {
+            this.providers.remove(DEFAULT_PROVIDER_KEY);
+        } else {
+            setProvider(DEFAULT_PROVIDER_KEY, provider);
+        }
```

Comment:
I think we need to write lock at the start of this method. An interleaving thread could alter the provider between any of these operations.

I think this means you'll have to change the locking to be able to cal [the overload](https://github.com/open-feature/java-sdk/pull/388/files#diff-0bd159679b99d067ac5064dcff6f517c1aed1ef3d6023499260568c67dfdc035R99) from this method (since you won't be able to lock twice on the same call stack, I think, unless that's supported magically somehow).

Reply:
So my thought was that the concurrent hash map should solve that? Both remove and the underlying set should be safe, as far as I understand.

Reply:
Ya, I think that part is safe. As written, I believe 2 threads could enter this method at the same time, and I'm worried could cause some races because the entire method is not atomic/locked. Perhaps I'm being overly cautious.

Reply:
This might be easy to fix with the suggestion here [1] : to have a clear separation between default provider & named providers 

[1] - https://github.com/open-feature/java-sdk/pull/388/files#r1185472185

Reply:
If we're going to have lock before writing into this map, `concurrentHashMap` seems to be bit overkill and we can use `HashMap` itself.

Reply:
Alright. I've accepted @thiyagu06's great suggestion about separating out the default case. The logic is much clearer now, but I think it's possible for someone to change the default provider out from someone. That said, I don't think this is a big deal and I'm inclined to leave it. Please let me know if that concerns you.

Reply:
current implementation looks good for me. btw it's a suggestion from @Kavindu-Dodan 😃